src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,99 @@
+﻿
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeQuality;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeStructFieldsWritable
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpMakeStructFieldsWritableDiagnosticAnalyzer : AbstractCodeQualityDiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor s_diagnosticDescriptor = CreateDescriptor(
+            IDEDiagnosticIds.MakeStructFieldsWritable,
+            new LocalizableResourceString(nameof(FeaturesResources.Make_readonly_fields_writable), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            new LocalizableResourceString(nameof(FeaturesResources.Make_readonly_fields_writable), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        public CSharpMakeStructFieldsWritableDiagnosticAnalyzer()
+            : base(ImmutableArray.Create(s_diagnosticDescriptor), GeneratedCodeAnalysisFlags.ReportDiagnostics)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        {
+            return DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => true;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext
+                => CompilationAnalyzer.CreateAndRegisterActions(compilationStartContext));
+        }
+
+        private sealed class CompilationAnalyzer
+        {
+            private bool _hasReadonlyField = false;
+            private bool _hasTypeInstanceAssigment = false;
+            private SyntaxNode _typeInstanceAssigmentNode;
+
+            public static void CreateAndRegisterActions(CompilationStartAnalysisContext compilationStartContext)
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // We are only interested in struct declarations
+                    var namedTypeSymbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                    if (namedTypeSymbol.TypeKind != TypeKind.Struct) return;
+
+                    var compilationAnalyzer = new CompilationAnalyzer();
```

Comment:
>CompilationAnalyzer [](start = 50, length = 19)

This should be renamed to `SymbolAnalyzer` as it is instantiated once per each struct type, not just once per compilation.

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/32063/commits/24d990cf937d249410387b5428ea6b23258e783b](https://github.com/dotnet/roslyn/pull/32063/commits/24d990cf937d249410387b5428ea6b23258e783b)